toolchest_client/tools/tool.py

```diff
@@ -89,70 +90,112 @@ def _prepare_inputs(self):
-                sanitized_args_list.append(arg)
-                num_args_remaining_after_tag = tool_arg_whitelist[minimal_tag]
+                sanitized_args.append(arg)
+                num_args_remaining_after_tag = whitelist[minimal_tag]
+                # if the arg is a tag in the dangerlist, note so can adjust accordingly
+                if minimal_tag in dangerlist:
+                    dangerous_args.append(arg)
             elif num_args_remaining_after_tag == VARIABLE_ARGS:
                 # if previous tag has additional args (unknown/variable amount),
                 # append args until another tag is found
                 # TODO: filter out non-escaped bash command-line characters
                 # TODO: handle variable arguments better; this allows passing of undesired args
-                sanitized_args_list.append(arg)
+                sanitized_args.append(arg)
                 if tag_in_whitelist:
-                    num_args_remaining_after_tag = tool_arg_whitelist[minimal_tag]
+                    num_args_remaining_after_tag = whitelist[minimal_tag]
             elif num_args_remaining_after_tag > 0:
                 # append remaining args if previous tag has additional args
                 # TODO: filter out non-escaped bash command-line characters
-                sanitized_args_list.append(arg)
+                sanitized_args.append(arg)
                 num_args_remaining_after_tag -= 1
-            # if no tag found, skip args until a tag is found
-
-        sanitized_args = " ".join(sanitized_args_list)
+            else:
+                # Instead of stopping and throwing an error immediately, we wait and collect all undesired args
+                if minimal_tag in blacklist:
+                    blacklisted_args.append(arg)
+                else:
+                    unknown_args.append(arg)
+
+        if unknown_args or blacklisted_args:
+            print("Non-allowed arguments found in tool_args:")
+            print(
+                f"Blacklisted arguments (these are known to cause Toolchest to fail): \
+{blacklisted_args if blacklisted_args else '(none)'}"
+            )
+            print(
+                f"Unknown arguments (these are not yet validated for use with Toolchest – please contact us!): \
+{unknown_args if unknown_args else '(none)'}"
+            )
+            raise ValueError("Unknown or blacklisted arguments present in tool_args. See above for details.")
+
+        if dangerous_args:
+            print("WARNING: dangerous arguments found in tool_args. This disables validation and parallelization!")
+            print(f"Dangerous arguments: {dangerous_args}")
+            # Disable parallelization, validation, and revert to plain compressed output
+            self.output_validation_enabled = False
+            self.parallel_enabled = False
+            self.output_is_directory = True
+            self.output_type = OutputType.GZ_TAR
```

Comment:
Should `self.output_type` have a default value (e.g., `OutputType.FLAT_TEXT`) in `Tool.__init__()`, for backwards compatibility? 

Unless `self.output_type` is explicitly defined for a tool's `SpecificTool.__init__()` function (which it probably should for every tool, just for clarity), it doesn't appear to be assigned a value explicitly unless `dangerous_args` are detected. This could pose a problem for any unpacking logic later.

Reply:
Oooh yes, good catch

Reply:
Added a default value of `OutputType.FLAT_TEXT`